### PR TITLE
Object name operator

### LINKED ISSUE #615: object_name_linter catches nonalphanumeric operator names
```
 writeLines("
`%++%` <- function(x, y) x+y
", tmp <- tempfile())

lintr::lint(tmp, lintr::object_name_linter())
# /tmp/Rtmp0dwFUb/file7f2d15642314:2:1: style: Variable and function name style should be snake_case.
# `%++%` <- function(x, y) x+y
# ^~~~~~
```

Seems like that's not possible to put into snake case...

### PR DESCRIPTION
**Pull request by MichaelChirico**
Closes #615

### CODE REVIEW

File: R/object_name_linters.R [hunk 1]
```diff
@@ -46,14 +45,16 @@ object_name_linter <- function(styles = "snake_case") {
...
-    assignments <- xml2::xml_find_all(x, xpath)
+    assignments <- xml2::xml_find_all(xml, xpath)
 
     # Retrieve assigned name
-    nms <- strip_names(
-      as.character(xml2::xml_find_first(assignments, "./text()")))
+    nms <- xml2::xml_text(assignments, "./text()")
+
+    # exclude operators if they're only symbols (e.g. %+%), #615
+    nms <- strip_names(grep("^[`'\"]%[^[:alnum:]]+%[`'\"]$", nms, invert = TRUE, value = TRUE))
```
Comment (MichaelChirico) [inline R/object_name_linters.R:54]: I don't know how to cleanly do this grep-filter with `rex`, do I have to use `!re_matches`?

**Comment by MichaelChirico** [comment]
Will add tests before merging

**Comment by AshesITR** [comment]
Wouldn't it be simpler to add a new name style, `operator_symbols` that allows `%[^[:alnum:]]+%` and add that to the default (i.e. `object_name_linter(style = c("snake_case", "operator_symbols"))`?

This has the benefit of not hard-wiring the exception so they can be disabled if desired.

**Comment by MichaelChirico** [comment]
great idea. much better.

File: R/object_name_linters.R (hunk 1 above)
Comment (MichaelChirico) [reply R/object_name_linters.R:54]: no longer needed

File: R/object_name_linters.R [hunk 2]
```diff
@@ -275,11 +273,12 @@ loweralnum <- rex(one_of(lower, digit))
 upperalnum <- rex(one_of(upper, digit))
 
 style_regexes <- list(
-  "CamelCase" = rex(start, maybe("."), upper, zero_or_more(alnum), end),
-  "camelCase" = rex(start, maybe("."), lower, zero_or_more(alnum), end),
-  "snake_case"     = rex(start, maybe("."), some_of(lower, digit), any_of("_", lower, digit), end),
-  "SNAKE_CASE"     = rex(start, maybe("."), some_of(upper, digit), any_of("_", upper, digit), end),
-  "dotted.case"    = rex(start, maybe("."), one_or_more(loweralnum), zero_or_more(dot, one_or_more(loweralnum)), end),
+  "symbols"     = rex(start, maybe("."), zero_or_more(none_of(alnum)), end),
```
Comment (AshesITR) [inline R/object_name_linters.R:276]: This allows a single dot and empty names. `rex(start, one_or_more(none_of(alnum)), end)` sounds like a cleaner approach.

File: tests/testthat/test-object_name_linter.R [hunk 3]
```diff
@@ -3,38 +3,38 @@ context("object_name_linter")
...
+  #                                                                       symbl   UpC   lowC   snake  SNAKE  dot    alllow  ALLUP
+  expect_equivalent(lapply(styles, check_style, nms = "IHave1Cat"    ), c(FALSE,  TRUE, FALSE, FALSE, FALSE, FALSE,  FALSE, FALSE))
+  expect_equivalent(lapply(styles, check_style, nms = "iHave1Cat"    ), c(FALSE, FALSE,  TRUE, FALSE, FALSE, FALSE,  FALSE, FALSE))
+  expect_equivalent(lapply(styles, check_style, nms = "i_have_1_cat" ), c(FALSE, FALSE, FALSE,  TRUE, FALSE, FALSE,  FALSE, FALSE))
+  expect_equivalent(lapply(styles, check_style, nms = "I_HAVE_1_CAT" ), c(FALSE, FALSE, FALSE, FALSE,  TRUE, FALSE,  FALSE, FALSE))
+  expect_equivalent(lapply(styles, check_style, nms = "i.have.1.cat" ), c(FALSE, FALSE, FALSE, FALSE, FALSE,  TRUE,  FALSE, FALSE))
+  expect_equivalent(lapply(styles, check_style, nms = "ihave1cat"    ), c(FALSE, FALSE,  TRUE,  TRUE, FALSE,  TRUE,   TRUE, FALSE))
+  expect_equivalent(lapply(styles, check_style, nms = "IHAVE1CAT"    ), c(FALSE,  TRUE, FALSE, FALSE,  TRUE, FALSE,  FALSE,  TRUE))
+  expect_equivalent(lapply(styles, check_style, nms = "I.HAVE_ONECAT"), c(FALSE, FALSE, FALSE, FALSE, FALSE, FALSE,  FALSE, FALSE))
 })
```
Comment (AshesITR) [inline tests/testthat/test-object_name_linter.R:38]: Can we add some positive test cases for symbols?
Especially `"."` would be an interesting edge case (Matching only for the style symbols although all others start with `maybe(".")`)

**Review by AshesITR — CHANGES_REQUESTED** [review]

File: R/object_name_linters.R (hunk 2 above)
Comment (MichaelChirico) [reply R/object_name_linters.R:276]: i thought of that but empty names won't parse:

    `` = 2
    # Error: attempt to use zero-length variable name
    assign("", 2) # same

but point taken that it's cleaner

**Review by AshesITR — APPROVED** [review]

**Review by AshesITR — APPROVED** [review]
LGTM, thanks!
